docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,213 @@
+  ]
+}'
+```
+
+The profile definition in the payload is an array of modules.
+Each module could be a firmware, software or configuration.
+The modules are installed/applied in the order in which they are defined in the profile definition.
+There is no restriction on the order of modules in a profile and hence can be defined in any preferred order.
+For example, additional software can be installed or configurations updated before the firmware is updated.
+
+The `"include"` field is optional and the value is true, by default.
+It can be used to exclude already installed modules when the same profile is reapplied on a device after a partial failure.
+
+When the profile is applied, `tedge-agent` is free to group modules of the same kind that are listed sequentially,
+to optimize the execution of that operation.
+For example, if 3 software modules are defined sequentially, the agent could group them into an `updateList`
+so that they can be installed in one-go using the `update-list` API of software plugins.
+
+## Tedge agent handling device profile commands
+
+The `tedge-agent` handless `device_profile` commands using the workflow definition at `/etc/tedge/operations/device_profile.toml`.
+This workflow definition handles each module type using sub-operation workflows defined for that type.
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+### On success
+
+Once all the modules in the profile are successfully installed, the successful status is published
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "successful",
+  ... // Other input fields
+}'
+```
+
+...and the current applied device profile information is updated by publishing the same to the capability topic as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile' '{
+  "name": "prod-profile",
+  "version: "v2"
+}'
+```
+
+### On failure
+
+On failure, the `device_profile` operation is aborted at the module that caused the failure.
+The remaining modules in the profile are not installed and no attempt is made to rollback the already installed modules either.
+If the sub-operations support rollbacks at the sub-operation level, it is performed for the failed module.
+
+For example, if a profile includes a firmware, 2 software and 1 configuration updates in that sequence,
+if the failure happens during the second software update, no rollback is performed at the overall `device_profile` operation level,
+or even for that failed software update, unless the `software_update` workflow for that software type supports rollbacks.
```

Comment:
>  unless the `software_update` workflow for that software type supports rollbacks.

What's the responsibility of the device profile workflow here?
Who triggers the rollback?

- Is this done by the sub-operation before reporting the failure?
- Or is this done by the device profile workflow? How then?

Reply:
In my mind, the `device_profile` workflow has no role in rollbacks. If it had, it should have done it at the entire profile level, which I don't think is feasible. Since rollbacks are typically handled at the individual operation workflow level (e.g: firmware update), it is left to that suboperation to rollback the failed module, if it can handle that.

The solution in my mind to the partially applied profile issue is to retry the same profile from the point of failure. In the current proposal it was left to the user to exclude all the already applied modules and resume from the failed module. But, as you suggested, we can make the agent maintain that state and automatically resume from the failed module without any user intervention.

Reply:
Added the firmware example as well.